src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts

```diff
@@ -0,0 +1,320 @@
+        this.log.debug(`initializing ${executor.name}`);
+
+        // Create root working directory after executor is ready
+        const nniRootDir: string = executor.joinPath(executor.getTempPath(), 'nni');
+        await executor.createFolder(executor.getRemoteExperimentRootDir(getExperimentId()));
+
+        // the directory to store temp scripts in remote machine
+        const remoteGpuScriptCollectorDir: string = executor.getRemoteScriptsPath(getExperimentId());
+
+        // clean up previous result.
+        await executor.createFolder(remoteGpuScriptCollectorDir, true);
+        await executor.allowPermission(true, nniRootDir);
+    }
+
+    public async refreshEnvironmentsStatus(environments: EnvironmentInformation[]): Promise<void> {
+        environments.forEach(async (environment) => {
+            const executor = await this.getExecutor(environment.id);
+            const jobpidPath: string = `${environment.runnerWorkingFolder}/pid`;
+            const trialReturnCodeFilePath: string = `${environment.runnerWorkingFolder}/code`;
+            if (fs.existsSync(jobpidPath)) {
+                /* eslint-disable require-atomic-updates */
+                try {
+                    const isAlive = await executor.isProcessAlive(jobpidPath);
+                    // if the process of jobpid is not alive any more
+                    if (!isAlive) {
+                        this.log.info(`pid in ${jobpidPath} is not alive!`);
+                        if (fs.existsSync(trialReturnCodeFilePath)) {
+                            const trialReturnCode: string = await executor.getRemoteFileContent(trialReturnCodeFilePath);
+                            const match: RegExpMatchArray | null = trialReturnCode.trim()
+                                .match(/^-?(\d+)\s+(\d+)$/);
+                            if (match !== null) {
+                                const { 1: code } = match;
+                                // Update trial job's status based on result code
+                                if (parseInt(code, 10) === 0) {
+                                    environment.setStatus('SUCCEEDED');
+                                } else {
+                                    environment.setStatus('FAILED');
+                                }
+                                this.releaseEnvironmentResource(environment);
+                            }
+                        }
+                    }
+                } catch (error) {
+                    this.releaseEnvironmentResource(environment);
+                    this.log.error(`Update job status exception, error is ${error.message}`);
+                }
+            }
+        });
+    }
+
+    /**
+     * If a environment is finished, release the connection resource
+     * @param environment remote machine environment job detail
+     */
+    public releaseEnvironmentResource(environment: EnvironmentInformation): void {
```

Comment:
private function? If it's proactive stopping, kill the trial runner.

Reply:
updated